virtcontainers/netns_watcher/netns_watcher.c

```diff
@@ -0,0 +1,1251 @@
+				addr, iface_list[idx].name);
+		return;
+	}
+
+	prev->next = tmp->next;
+
+free_node:
+	free(tmp->addr);
+	free(tmp);
+}
+
+/* 
+ * Validate the parameters and fill the appropriate structure.
+ *
+ * pod-id: Pod identifier related to the network namespace this watcher
+ * has to listen to.
+ * runtime-path: Runtime path is needed to call into it whenever a
+ * change in the network namespace is detected.
+ * netns-path: Network namespace path that needs to be monitored.
+ *
+ * It returns 0 if the parameters are valid, with a watcher_params structure
+ * properly filled. Otherwise, it returns -1 with a null watcher_params
+ * structure.
+ */
+int valid_params(int argc, const char **argv, struct watcher_params *params)
+{
+	struct option prog_opts[] = {
+		{"debug", no_argument, 0, 'd'},
+		{"help", no_argument, 0, 'h'},
+		{"netns-path", required_argument, 0, 'n'},
+		{"pod-id", required_argument, 0, 'p'},
+		{"runtime-path", required_argument, 0, 'r'},
+		{"version", no_argument, 0, 'v'},
+		{ 0, 0, 0, 0},
+	};
+
+	int c;
+	while ((c = getopt_long(argc, (char **)argv, "dhn:p:r:v",
+					prog_opts, NULL)) != -1) {
+		switch (c) {
+		case 'd':
+			debug = true;
+			break;
+		case 'h':
+			print_usage();
+			exit(EXIT_SUCCESS);
+		case 'n':
+			params->netns_path = strdup(optarg);
+			break;
+		case 'p':
+			params->pod_id = strdup(optarg);
+			break;
+		case 'r':
+			params->runtime_path = strdup(optarg);
+			break;
+		case 'v':
+			print_version();
+			exit(EXIT_SUCCESS);
+		default:
+			print_usage();
+			exit(EXIT_FAILURE);
+		}
+	}
+
+	if (!params->netns_path) {
+		printf("Missing network namespace path\n");
+		goto err;
+
+	}
+
+	if (!params->pod_id) {
+		printf("Missing pod ID\n");
+		goto err;
+	}
+
+	if (!params->runtime_path) {
+		printf("Missing runtime path\n");
+		goto err;
+	}
+
+	return 0;
+
+err:
+	free_watcher_params(params);
+	return -1;
+}
+
+/* 
+ * Enter the network namespace.
+ *
+ * Entering the provided network namespace since the point of this binary is
+ * to monitor any network change happening inside a specific network namespace.
+ */
+int enter_netns(const char *netns_path)
+{
+	int fd;
+
+	fd = open(netns_path, O_RDONLY);
+	if (fd == -1) {
+		printf("Failed opening network ns %s: %s\n",
+				netns_path, strerror(errno));
+		return -1;
+	}
+
+	if (setns(fd, 0) == -1) {
+		printf("Failed to join network ns %s: %s\n",
+				netns_path, strerror(errno));
+		return -1;
+	}
+
+	return 0;
+}
+
+/*
+ * Open the netlink socket with appropriate parameters, and return the file
+ * descriptor. It will listen for any message related to a network interface
+ * change (add/delete/update), an IP address (IPv4) change (add/delete/update)
+ * or a route (IPv4) change (add/delete/update).
+ */
+int open_netlink()
+{
+	int sock;
+	struct sockaddr_nl sa;
+
+	memset(&sa, 0, sizeof(sa));
+
+	sa.nl_family = AF_NETLINK;
+	sa.nl_pid = getpid();
+	sa.nl_groups = RTMGRP_LINK |
+		RTMGRP_IPV4_IFADDR |
```

Comment:
hard-coded IPv4. Is this intentional?

Reply:
Yes, as a first step, I want to make sure we handle properly ipv4 only.